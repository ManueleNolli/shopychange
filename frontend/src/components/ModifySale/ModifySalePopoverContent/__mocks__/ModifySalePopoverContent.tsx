import { divWithoutChildrenMock } from '../../../../../mocksUtils'

const ModifySalePopoverContentMock = () =>
  divWithoutChildrenMock('ModifySalePopoverContent')
export default ModifySalePopoverContentMock
