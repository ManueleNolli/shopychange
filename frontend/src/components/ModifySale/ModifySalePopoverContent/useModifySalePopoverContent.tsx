import { useState } from 'react'

import { useToast } from '@chakra-ui/react'
import {
  modifySale,
  waitTransaction,
} from '../../../services/BlockchainService'

import { useUserContext } from '../../../context/userContext'
import useLoading from '../../../hooks/useLoading'
import { ModifySalePopoverContentProps } from '../../../types/props/ModifySalePopoverContentProps'

export default function useModifySalePopoverContent({
  nft,
  onClose,
  onSucess,
  actualPrice,
}: ModifySalePopoverContentProps) {
  const toast = useToast()
  const { userAddress, blockchainNetworkId } = useUserContext()
  const { isLoading, setIsLoading } = useLoading()
  const [newPrice, setnewPrice] = useState<number>(actualPrice)
  const [diffPercentage, setDiffPercentage] = useState<number>(0)

  const onPriceChange = (value: string) => {
    if (value) {
      const valueNum = parseFloat(value)
      setDiffPercentage(((valueNum - actualPrice) / actualPrice) * 100)
      setnewPrice(valueNum)
    }
  }

  const modify = async () => {
    if (!userAddress || !blockchainNetworkId) return
    if (newPrice && newPrice != actualPrice) {
      setIsLoading(true)
      return modifySale(
        blockchainNetworkId,
        nft.contractAddress,
        nft.tokenId,
        newPrice
      )
        .then(waitTransaction)
        .then(() => {
          setIsLoading(false)
          onSucess()
          onClose()
        }) // eslint-disable-next-line @typescript-eslint/no-explicit-any
        .catch((err: any) => {
          // First phrase of the error message is enough
          const errorMessage = err.message.split('.')[0]
          toast({
            variant: 'solid',
            title: 'Transaction failed',
            description: errorMessage,
            status: 'warning',
            duration: 9000,
            isClosable: true,
          })
          setIsLoading(false)
        })
    }
  }

  return {
    isLoading,
    newPrice,
    diffPercentage,
    onPriceChange,
    modify,
  }
}
