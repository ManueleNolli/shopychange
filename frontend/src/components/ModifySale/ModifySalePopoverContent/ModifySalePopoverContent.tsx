import React from 'react'

import {
  ButtonGroup,
  Center,
  Container,
  Flex,
  Heading,
  NumberInput,
  NumberInputField,
  Spacer,
  Table,
  TableContainer,
  Tbody,
  Td,
  Text,
  Th,
  Thead,
  Tr,
} from '@chakra-ui/react'

import LoadingButton from '../../LoadingButton/LoadingButton'
import ShadowButton from '../../ShadowButton/ShadowButton'

import { FaLongArrowAltRight } from 'react-icons/fa'

import { ModifySalePopoverContentProps } from '../../../types/props/ModifySalePopoverContentProps'

import { colors } from '../../../styles/Colors'
import PriceDistrubution from '../../PriceDistribition/PriceDistribution'
import useModifySalePopoverContent from './useModifySalePopoverContent'

export default function ModifySalePopoverContent({
  nft,
  onClose,
  onSucess,
  actualPrice,
}: ModifySalePopoverContentProps) {
  const { isLoading, newPrice, diffPercentage, onPriceChange, modify } =
    useModifySalePopoverContent({
      nft,
      onClose,
      onSucess,
      actualPrice,
    })

  return (
    <Flex direction="column" flex="1" p="1vh">
      <Spacer />

      <Container flex="1" centerContent>
        <Heading size="lg" textAlign={'center'}>
          Are you sure you want to modify <Text as="i">{nft.name}</Text>?
        </Heading>
      </Container>

      <>
        <Spacer />
        <Center>
          <TableContainer>
            <Table variant="unstyled">
              <Thead>
                <Tr>
                  <Th>
                    <Text textAlign={'center'}>Actual Price</Text>
                  </Th>
                  <Th></Th>
                  <Th>
                    <Text textAlign={'center'}>New Price</Text>
                  </Th>
                </Tr>
              </Thead>
              <Tbody>
                <Tr>
                  <Td>
                    <Center dir="column">
                      <NumberInput isDisabled value={actualPrice}>
                        <NumberInputField textAlign="center" />
                      </NumberInput>
                      <Text pl="2" color={'gray.500'}>
                        eth
                      </Text>
                    </Center>
                  </Td>
                  <Td>
                    <Text
                      as="b"
                      color={diffPercentage >= 0 ? 'green' : 'red'}
                      textAlign={'center'}
                    >
                      {diffPercentage.toFixed(2)} %
                    </Text>
                    <Center>
                      <FaLongArrowAltRight
                        size={'2em'}
                        color={diffPercentage >= 0 ? 'green' : 'red'}
                      />
                    </Center>
                  </Td>
                  <Td>
                    <Center dir="column">
                      <NumberInput
                        defaultValue={actualPrice}
                        min={0}
                        step={0.01}
                        onChange={onPriceChange}
                      >
                        <NumberInputField
                          textAlign="center"
                          data-testid="price-input"
                        />
                      </NumberInput>
                      <Text pl="2" color={'gray.500'}>
                        eth
                      </Text>
                    </Center>
                  </Td>
                </Tr>
              </Tbody>
            </Table>
          </TableContainer>
        </Center>
      </>
      <Spacer />

      <PriceDistrubution
        price={newPrice}
        isOwner={true}
        address={nft.contractAddress}
        tokenId={nft.tokenId}
      />

      <Spacer />

      <Center flex="1">
        <ButtonGroup spacing="10">
          <LoadingButton isLoading={isLoading} data-testid="loading-button">
            <ShadowButton
              label={'Modify price'}
              bgColor={`${colors.modifyChakraColor}.400`}
              textColor={'white'}
              shadowColor={colors.modifyShadowColor}
              hoverColor={`${colors.modifyChakraColor}.500`}
              focusColor={`${colors.modifyChakraColor}.500`}
              onClick={modify}
            />
          </LoadingButton>
          <ShadowButton
            label={'Cancel'}
            bgColor={`${colors.cancelChakraColor}.500`}
            textColor={'white'}
            shadowColor={colors.cancelShadowColor}
            hoverColor={`${colors.cancelChakraColor}.600`}
            focusColor={`${colors.cancelChakraColor}.600`}
            onClick={onClose}
          />
        </ButtonGroup>
      </Center>
      <Spacer />
    </Flex>
  )
}
