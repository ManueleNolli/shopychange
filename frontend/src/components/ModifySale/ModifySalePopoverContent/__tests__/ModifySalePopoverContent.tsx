import React from 'react'
import { render } from '@testing-library/react'
import ModifySalePopoverContent from '../ModifySalePopoverContent'
import useModifySalePopoverContent from '../useModifySalePopoverContent'

const mockProps = {
  nft: {
    contractAddress: '0x0000000000',
    tokenId: 0,
    image:
      'https://ipfs.io/ipfs/bafybeietntip4sq3xmuwiz3s43yb3klrpa3okgtg5deysmbd5hbizhrb2q',
    name: 'Test NFT',
  },
  onClose: jest.fn(),
  actualPrice: 0.4,
}
jest.mock('../useModifySalePopoverContent', () => jest.fn())

jest.mock('../../../PriceDistribition/PriceDistribution')

const onUpdate = jest.fn()

describe('SellPopoverContent', () => {
  test('renders correctly diff > 0', () => {
    ;(useModifySalePopoverContent as jest.Mock).mockReturnValue({
      isLoading: false,
      newPrice: 10,
      diffPercentage: 50,
      onPriceChange: jest.fn(),
      modify: jest.fn(),
    })
    const { container } = render(
      <ModifySalePopoverContent
        nft={mockProps.nft}
        onClose={mockProps.onClose}
        actualPrice={mockProps.actualPrice}
        onSucess={onUpdate}
      />
    )

    expect(container).toMatchSnapshot()
  })

  test('renders correctly diff < 0', () => {
    ;(useModifySalePopoverContent as jest.Mock).mockReturnValue({
      isLoading: false,
      newPrice: 10,
      diffPercentage: -50,
      onPriceChange: jest.fn(),
      modify: jest.fn(),
    })
    const { container } = render(
      <ModifySalePopoverContent
        nft={mockProps.nft}
        onClose={mockProps.onClose}
        actualPrice={mockProps.actualPrice}
        onSucess={onUpdate}
      />
    )

    expect(container).toMatchSnapshot()
  })
})
