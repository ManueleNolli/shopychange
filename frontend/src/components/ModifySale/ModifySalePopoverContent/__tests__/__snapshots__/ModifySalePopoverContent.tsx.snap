// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SellPopoverContent renders correctly diff < 0 1`] = `
<div>
  <div
    class="css-1wjy6p3"
  >
    <div
      class="css-17xejub"
    />
    <div
      class="chakra-container css-11uhf22"
    >
      <h2
        class="chakra-heading css-xi606m"
      >
        Are you sure you want to modify 
        <i
          class="chakra-text css-0"
        >
          Test NFT
        </i>
        ?
      </h2>
    </div>
    <div
      class="css-17xejub"
    />
    <div
      class="css-gmuwbf"
    >
      <div
        class="chakra-table__container css-zipzvv"
      >
        <table
          class="chakra-table css-0"
        >
          <thead
            class="css-0"
          >
            <tr
              class="css-0"
            >
              <th
                class="css-0"
              >
                <p
                  class="chakra-text css-xi606m"
                >
                  Actual Price
                </p>
              </th>
              <th
                class="css-0"
              />
              <th
                class="css-0"
              >
                <p
                  class="chakra-text css-xi606m"
                >
                  New Price
                </p>
              </th>
            </tr>
          </thead>
          <tbody
            class="css-0"
          >
            <tr
              class="css-0"
            >
              <td
                class="css-0"
              >
                <div
                  class="css-gmuwbf"
                  dir="column"
                >
                  <div
                    class="chakra-numberinput css-e69dqy"
                    value="0.4"
                  >
                    <input
                      aria-valuemax="9007199254740991"
                      aria-valuemin="-9007199254740991"
                      aria-valuenow="0.4"
                      aria-valuetext="0.4"
                      autocomplete="off"
                      autocorrect="off"
                      class="chakra-numberinput__field css-fzwid6"
                      disabled=""
                      inputmode="decimal"
                      pattern="[0-9]*(.[0-9]+)?"
                      role="spinbutton"
                      type="text"
                      value="0.4"
                    />
                  </div>
                  <p
                    class="chakra-text css-td3r9l"
                  >
                    eth
                  </p>
                </div>
              </td>
              <td
                class="css-0"
              >
                <b
                  class="chakra-text css-1plnjxl"
                >
                  -50.00
                   %
                </b>
                <div
                  class="css-gmuwbf"
                >
                  <svg
                    color="red"
                    fill="currentColor"
                    height="2em"
                    stroke="currentColor"
                    stroke-width="0"
                    style="color: red;"
                    viewBox="0 0 448 512"
                    width="2em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M313.941 216H12c-6.627 0-12 5.373-12 12v56c0 6.627 5.373 12 12 12h301.941v46.059c0 21.382 25.851 32.09 40.971 16.971l86.059-86.059c9.373-9.373 9.373-24.569 0-33.941l-86.059-86.059c-15.119-15.119-40.971-4.411-40.971 16.971V216z"
                    />
                  </svg>
                </div>
              </td>
              <td
                class="css-0"
              >
                <div
                  class="css-gmuwbf"
                  dir="column"
                >
                  <div
                    class="chakra-numberinput css-e69dqy"
                  >
                    <input
                      aria-valuemax="9007199254740991"
                      aria-valuemin="0"
                      aria-valuenow="0.4"
                      aria-valuetext="0.40"
                      autocomplete="off"
                      autocorrect="off"
                      class="chakra-numberinput__field css-fzwid6"
                      data-testid="price-input"
                      inputmode="decimal"
                      pattern="[0-9]*(.[0-9]+)?"
                      role="spinbutton"
                      type="text"
                      value="0.40"
                    />
                  </div>
                  <p
                    class="chakra-text css-td3r9l"
                  >
                    eth
                  </p>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="css-17xejub"
    />
    <div
      data-testid="PriceDistribution"
    />
    <div
      class="css-17xejub"
    />
    <div
      class="css-1en42kf"
    >
      <div
        class="chakra-button__group css-s2zvk0"
        data-orientation="horizontal"
        role="group"
      >
        <button
          class="chakra-button css-ha7ah3"
          data-testid="onClick-button"
          style="white-space: initial;"
          type="button"
        >
          Modify price
        </button>
        <button
          class="chakra-button css-lnz0i6"
          data-testid="onClick-button"
          style="white-space: initial;"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
    <div
      class="css-17xejub"
    />
  </div>
</div>
`;

exports[`SellPopoverContent renders correctly diff > 0 1`] = `
<div>
  <div
    class="css-1wjy6p3"
  >
    <div
      class="css-17xejub"
    />
    <div
      class="chakra-container css-11uhf22"
    >
      <h2
        class="chakra-heading css-xi606m"
      >
        Are you sure you want to modify 
        <i
          class="chakra-text css-0"
        >
          Test NFT
        </i>
        ?
      </h2>
    </div>
    <div
      class="css-17xejub"
    />
    <div
      class="css-gmuwbf"
    >
      <div
        class="chakra-table__container css-zipzvv"
      >
        <table
          class="chakra-table css-0"
        >
          <thead
            class="css-0"
          >
            <tr
              class="css-0"
            >
              <th
                class="css-0"
              >
                <p
                  class="chakra-text css-xi606m"
                >
                  Actual Price
                </p>
              </th>
              <th
                class="css-0"
              />
              <th
                class="css-0"
              >
                <p
                  class="chakra-text css-xi606m"
                >
                  New Price
                </p>
              </th>
            </tr>
          </thead>
          <tbody
            class="css-0"
          >
            <tr
              class="css-0"
            >
              <td
                class="css-0"
              >
                <div
                  class="css-gmuwbf"
                  dir="column"
                >
                  <div
                    class="chakra-numberinput css-e69dqy"
                    value="0.4"
                  >
                    <input
                      aria-valuemax="9007199254740991"
                      aria-valuemin="-9007199254740991"
                      aria-valuenow="0.4"
                      aria-valuetext="0.4"
                      autocomplete="off"
                      autocorrect="off"
                      class="chakra-numberinput__field css-fzwid6"
                      disabled=""
                      inputmode="decimal"
                      pattern="[0-9]*(.[0-9]+)?"
                      role="spinbutton"
                      type="text"
                      value="0.4"
                    />
                  </div>
                  <p
                    class="chakra-text css-td3r9l"
                  >
                    eth
                  </p>
                </div>
              </td>
              <td
                class="css-0"
              >
                <b
                  class="chakra-text css-ggdrs2"
                >
                  50.00
                   %
                </b>
                <div
                  class="css-gmuwbf"
                >
                  <svg
                    color="green"
                    fill="currentColor"
                    height="2em"
                    stroke="currentColor"
                    stroke-width="0"
                    style="color: green;"
                    viewBox="0 0 448 512"
                    width="2em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M313.941 216H12c-6.627 0-12 5.373-12 12v56c0 6.627 5.373 12 12 12h301.941v46.059c0 21.382 25.851 32.09 40.971 16.971l86.059-86.059c9.373-9.373 9.373-24.569 0-33.941l-86.059-86.059c-15.119-15.119-40.971-4.411-40.971 16.971V216z"
                    />
                  </svg>
                </div>
              </td>
              <td
                class="css-0"
              >
                <div
                  class="css-gmuwbf"
                  dir="column"
                >
                  <div
                    class="chakra-numberinput css-e69dqy"
                  >
                    <input
                      aria-valuemax="9007199254740991"
                      aria-valuemin="0"
                      aria-valuenow="0.4"
                      aria-valuetext="0.40"
                      autocomplete="off"
                      autocorrect="off"
                      class="chakra-numberinput__field css-fzwid6"
                      data-testid="price-input"
                      inputmode="decimal"
                      pattern="[0-9]*(.[0-9]+)?"
                      role="spinbutton"
                      type="text"
                      value="0.40"
                    />
                  </div>
                  <p
                    class="chakra-text css-td3r9l"
                  >
                    eth
                  </p>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="css-17xejub"
    />
    <div
      data-testid="PriceDistribution"
    />
    <div
      class="css-17xejub"
    />
    <div
      class="css-1en42kf"
    >
      <div
        class="chakra-button__group css-s2zvk0"
        data-orientation="horizontal"
        role="group"
      >
        <button
          class="chakra-button css-ha7ah3"
          data-testid="onClick-button"
          style="white-space: initial;"
          type="button"
        >
          Modify price
        </button>
        <button
          class="chakra-button css-lnz0i6"
          data-testid="onClick-button"
          style="white-space: initial;"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
    <div
      class="css-17xejub"
    />
  </div>
</div>
`;
